### PR TITLE
Fixed Some bugs of code_mobject.py and Text_mobject.py

### PR DESCRIPTION
**Pull request by NavpreetDevpuri**
Fixed https://github.com/3b1b/manim/issues/1067
NOTE : SurroundingRectangle() only contains visible text.
```Python
class te(Scene):
    def construct(self):
        text1 = Text(" ab ", font="Consolas", size=2)
        rect1 = SurroundingRectangle(text1)
        text1.chars[1].set_color(GREEN)
        self.add(text1,rect1)
```
Output
te

```python
class bug1(Scene):
    def construct(self):
        text1 = Text("  ab\ncd", font="Consolas", size=2)
        text2 = Text("ab\ngh", font="Consolas", size=2)
        rect1 = SurroundingRectangle(text1)
        rect2 = SurroundingRectangle(text2)
        self.play(Transform(remove_invisible_chars(text1), remove_invisible_chars(text2)), Transform(rect1, rect2))
        self.wait()
```
Output 
bug1

Added new parameters `background_stroke_width` and `background_stroke_color`
```python
class bug2andbug3(Scene):
    def construct(self):
        helloworldc = Code(
            "helloworldc.c",
            tab_width=4,
            background_stroke_width=1,
            background_stroke_color=WHITE,
            insert_line_no=True,
            style=Code.styles_list[4],
            background="window",
            language="cpp",
        )
        helloworldcpp = Code(
            "helloworldcpp.cpp",
            tab_width=4,
            background_stroke_width=1,
            background_stroke_color=WHITE,
            insert_line_no=True,
            style=Code.styles_list[15],
            background="window",
            language="cpp",
        )
        helloworldc.move_to(np.array([-3.6, 0, 0]))
        helloworldcpp.move_to(np.array([3.1, 0, 0]))
        self.add(helloworldc,helloworldcpp)
```
Output
bug2andbug3

Updated Paragraph() and added new methods `set_line_to_initial_position()` and `set_all_lines_to_initial_positions()`
```python
class para(Scene):
    def construct(self):
        t = Paragraph('this is a awesome', 'paragraph', 'With \nNewlines', '\tWith Tabs', '  With Spaces',
                      'With Alignments',
                      'center', "left", "right")
        t.set_line_alignment("center", 7)
        t.set_line_alignment("right", 9)
        t.shift(2 * RIGHT)
        rect = SurroundingRectangle(t)
        self.add(t, rect)
        self.wait()
        self.play(t.set_all_lines_alignments, "right")
        self.play(t.set_line_to_initial_position, 4)
        self.play(t.set_all_lines_to_initial_positions)
        self.play(t.set_line_alignment, "center", 7)
        self.play(t.set_line_alignment, "right", 9)
        t.chars[0][5:7].set_color(GREEN)
        t[1][0:4].set_color(YELLOW)
        self.wait()
```
Output
ezgif-3-3dc23c79c5fa

### CODE REVIEW

**Comment by NavpreetDevpuri** [comment]
> What about trying rewriting `__getitem__()` to make `Text[i]` stand for all chars instead of using `.chars`.

i tried but don't work afficiently

**Comment by NavpreetDevpuri** [comment]
> Sorry, but I still think it is not the best solution for using `.chars[i]` to stand for all chars, and `Text[i]` stands for visible chars. This will make a big difference to people who has already used `Text`.
> I think it’s better not to change the structure or usage of `Text`.

25 days ago we added spaces before that `Text[i]` stands for visible chars
https://github.com/3b1b/manim/issues/1017

**Comment by NavpreetDevpuri** [comment]
> Sorry, but I still think it is not the best solution to use `.chars[i]` to stand for all chars, and `Text[i]` stands for visible chars. This will make a big difference to people who has already used `Text`.

How many of them already used `Text` ?
if they used they where unable to perform Transformation animations
and if they use they probibaly used t2c ..etc

**Comment by Tony031218** [comment]
> 25 days ago we added spaces, before that `Text[i]` stands for visible chars

No, before you create `.chars` `Text[i]` stands for visable chars.

**Comment by NavpreetDevpuri** [comment]
> and if they use they probibaly used t2c ..etc

Probibaly No one is using Text[] directly 
and if someone used it then he/she can easily debug it

**Comment by NavpreetDevpuri** [comment]
> I think it’s better not to change the structure or usage of `Text`.

what if all community thinks that then this manim is not possibly not made You can see almost no 3b1b videos scripts don't work because we improved manim

**Comment by Tony031218** [comment]
Many people I know are using `Text`. And will directly use `Text[i]` to represent the i-th of all characters, so as to modify individual characters. This will make it easier to use. 
I just give you some suggestions for improvement, if you do not accept, I can do nothing.

**Comment by NavpreetDevpuri** [comment]
let me think about it 
what you want to suggest ?
may be i can just leave Dots in Text and use remove_invisible_chars()

**Comment by Tony031218** [comment]
> Simple is better than complex.
>Special cases aren't special enough to break the rules.
This is the original sentence in the zen of python. I think adding `.chars` to represent all characters are complex and break the rules.

My suggestion is to still use `Text [i]` to represent the i-th of all characters. Because this will not break the previous usage.

**Comment by leotrs** [comment]
@NavpreetDevpuri Thanks for this! I really appreciate this!

A couple of things though. We have been changing some things around, and we won't be able to merge this since your code is meant for the old version of manim. Please, merge pull our master branch on top of this PR, resolve the merge conflicts, and make a new commit.

In particular, we have changed `manimlib` to `manim`. Also, we have changed `from manim.imports import *` to just `from manim import *`.

Does anyone know why this didn't trigger Travis?

**Comment by kilacoda** [comment]
> Does anyone know why this didn't trigger Travis?

Maybe because of the conflicts?

**Comment by leotrs** [comment]
> > Does anyone know why this didn't trigger Travis?
> 
> Maybe because of the conflicts?

Ah, fair.

@NavpreetDevpuri Please follow my instructions to resolve the merge conflicts on your local branch. If you get stuck, please post here (we do not yet have a good changelog of what has been going on in ManimCommunity...). Once those are done, a Travis-CI build should automatically be triggered to run some tests. As soon as the tests pass, we will be able to merge this.

**Comment by NavpreetDevpuri** [comment]
> @NavpreetDevpuri Thanks for this! I really appreciate this!
>
> A couple of things though. We have been changing some things around, and we won't be able to merge this since your code is meant for the old version of manim.

Fixed !

**Comment by NavpreetDevpuri** [comment]
i forked manim from 3b1b/manim may be thats why it creates conflicts
@eulertour please merge https://github.com/3b1b/manim/pull/1071 
after that i will fork from ManimCommunity/manim and create new PR

**Comment by leotrs** [comment]
@NavpreetDevpuri I'm not sure that merging 1071 into 3b1b is going to be sufficient. The point is that right now our fork and 3b1b are already diverging. The easiest path for us to merge this into our fork is by you cloning our fork, implementing your changes on that clone, and PRing from it (instead of PRing from a fork of 3b1b).

**Comment by PgBiel** [comment]
I much prefer if we keep `Text[i]` as it is because of the `str[i]` pattern in stdlib. For only visible characters you could create something else like `.vis_chars` or smth

**Comment by NavpreetDevpuri** [comment]
> I'd very much prefer we keep `Text[i]` as it is because of the `str[i]` pattern in stdlib. For only visible characters you could create something else like `.vis_chars` or smth


Now its fixed `Text[i]` stand for all chars

**Comment by NavpreetDevpuri** [comment]
```python
'''
Text is VGroup() of each characters
that mean you can use it like 
    Text[0:5] or Text.chars[0:5] to access first five characters 

Text[0:5] or Text or Text.chars[0:5] will create problems when using Transform() because of invisible characters 
so, before using Transform() remove invisible characters by using remove_invisible_chars()
for example self.play(Transform(remove_invisible_chars(text.chars[0:4]), remove_invisible_chars(text2.chars[0:2])))
'''

```

**Comment by NavpreetDevpuri** [comment]
> @NavpreetDevpuri I'm not sure that merging 1071 into 3b1b is going to be sufficient. The point is that right now our fork and 3b1b are already diverging. The easiest path for us to merge this into our fork is by you cloning our fork, implementing your changes on that clone, and PRing from it (instead of PRing from a fork of 3b1b).


https://github.com/NavpreetDevpuri/manim/pull/1
It still creates problems 

When 1071 is merge into 3b1b then i will delete my manim fork and after that i will fork from ManimCommunity/manim to my profile. will this create a clone of ManimCommunity/manim to my profile ?

**Comment by Tony031218** [comment]
> ```python
> 
> '''
> 
> Text is VGroup() of each characters
> 
> that mean you can use it like 
> 
>     Text[0:5] or Text.chars[0:5] to access first five characters 
> 
> 
> 
> Text[0:5] or Text or Text.chars[0:5] will create problems when using Transform() because of invisible characters 
> 
> so, before using Transform() remove invisible characters by using remove_invisible_chars()
> 
> for example self.play(Transform(remove_invisible_chars(text.chars[0:4]), remove_invisible_chars(text2.chars[0:2])))
> 
> '''
> 
> 
> 
> ```


This is good, but maybe you can apply a patch in Transform to automatically use the remove_invisible_chars function for Text.

**Comment by NavpreetDevpuri** [comment]
> This is good, but maybe you can apply a patch in Transform to automatically use the remove_invisible_chars function for Text.

how can be do that ?
i think there is no way to detect whether Transform is used for normal VGroup() or our Text's VGroup() or may be some other form of VGroup.

**Comment by NavpreetDevpuri** [comment]
May be we can create empty paths in svg file and after that we can remove empty paths in Transform method just before animation in that way we don't even need to create Dot() in submobjects

what you think ? is this ok or will this(empty paths) create problems in other aspects of manim ?

**Comment by NavpreetDevpuri** [comment]
> This is good, but maybe you can apply a patch in Transform to automatically use the remove_invisible_chars function for Text.

may be we can think about it in next PR

**Comment by leotrs** [comment]
@NavpreetDevpuri any news on this? There only remains the issue of documentation :)

**Comment by NavpreetDevpuri** [comment]
> @NavpreetDevpuri any news on this? There only remains the issue of documentation :)

i checked everything from my side. Its ready for merge.
but, may be @eulertour is checking [PR](https://github.com/3b1b/manim/pull/1071). Not merged till now. 

i added some documentations in python files as some comments 
let me know if it need more clear documentation. i think we don't have a clear documentations about manim till now.

**Comment by NavpreetDevpuri** [comment]
> Please add documentation. It'd also be nice if the documentation guidelines outlined in the Wiki are followed when you write it

now i see your comment 
let me check this

**Comment by NavpreetDevpuri** [comment]
> @NavpreetDevpuri any news on this? There only remains the issue of documentation :)

i was keeping track on original [PR]( https://github.com/3b1b/manim/pull/1071)
We will create new PR latter for documentation after this merge. 
Nowdays i just started an another [project](https://github.com/NavpreetDevpuri/tIedoC). I working on it.

**Comment by PgBiel** [comment]
It is in our guidelines to include documentation in your PR. You can take all the time you want for that, don't worry

**Comment by NavpreetDevpuri** [comment]
@PgBiel please review documentions of Code().
is it Ok ?

**Comment by NavpreetDevpuri** [comment]
This is my first documentation.

File: manimlib/mobject/svg/code_mobject.py [hunk 1]
```diff
@@ -1,52 +1,118 @@
...
+so, before using Transform() remove invisible characters by using remove_invisible_chars()
+for example self.play(Transform(remove_invisible_chars(Code.code.chars[0:2]), remove_invisible_chars(Code.code.chars[3][0:3])))
+or remove_invisible_chars(Code.code) or remove_invisible_chars(Code)
 '''
 
 
 class Code(VGroup):
+    """Class Code is used to display code with color highlighted.
+    Parameters
+    ----------
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:42]:
```suggestion
    """Class Code is used to display code with color highlighted.
    
    Parameters
    ----------
```

File: manimlib/mobject/svg/code_mobject.py [hunk 2]
```diff
@@ -1,52 +1,118 @@
...
+    line_no_buff : :class:`float`, optional (default is 0.4)
+        It defines space between line numbers and displayed code.
+    style : :class:`str`, optional (default is 'vim')
+        It defines style type of displayed code. You can see names of styles from Code.styles_list.
+    language : :class:`str`, optional (default is None, Which mean it will automatically detects the language)
+        It defines the language name of given code.
+        For Knowing more available options visit https://pygments.org/docs/lexers/  for 'aliases or short names'
+    generate_html_file : :class:`bool`, optional (default is False)
+        It defines whether to generate code highlighted html to folder assets/codes/generated_html_files.
+    Attributes (It have all Parameters as Attributes and some extra as following)
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:80]: the line must not contain anything other than "Attributes"
```suggestion
        It defines whether to generate code highlighted html to folder assets/codes/generated_html_files.
        
    Attributes
```

File: manimlib/mobject/svg/code_mobject.py [hunk 3]
```diff
@@ -245,6 +333,12 @@ def gen_code_json(self):
         # print(self.code_json)
 
     def correct_non_span(self, line_str):
+        """Function put text color to those strings that don't have one according to background_color of displayed code.
+        Arguments
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:337]: wrong name
```suggestion
        Parameters
```

File: manimlib/mobject/svg/code_mobject.py [hunk 4]
```diff
@@ -272,8 +366,23 @@ def correct_non_span(self, line_str):
         return line_str
 
 
-def hilite_me(code, lexer, options, style, linenos, divstyles):
-    lexer = lexer or 'python'
+def hilite_me(code, language, style, insert_line_no, divstyles, file_path):
+    """Function to highlight code from string to html.
+    Arguments
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:371]:
```suggestion

    Parameters
```

File: manimlib/mobject/svg/code_mobject.py [hunk 5]
```diff
@@ -283,18 +392,24 @@ def hilite_me(code, lexer, options, style, linenos, divstyles):
...
     return html
 
 
-def get_default_style():
-    return 'border:solid gray;border-width:.1em .1em .1em .8em;padding:.2em .6em;'
-
-
 def insert_line_numbers(html):
+    """Function put line numbers to html of highlighted code.
+    Arguments
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:408]:
```suggestion

    Parameters
```

File: manimlib/mobject/svg/text_mobject.py [hunk 6]
```diff
@@ -314,86 +307,180 @@ def text2svg(self):
...
                 last_line_num = setting.line_num
-            context.move_to(START_X + offset_x, START_Y + lsh*setting.line_num)
+            context.move_to(START_X + offset_x, START_Y + line_spacing * setting.line_num)
             context.show_text(text)
             offset_x += context.text_extents(text)[4]
+        surface.finish()
+        return file_name
+
+
+class TextWithBackground(Text):
```
Comment (PgBiel) [inline manimlib/mobject/svg/text_mobject.py:317]: docs?

File: manimlib/mobject/svg/text_mobject.py [hunk 7]
```diff
@@ -314,86 +307,180 @@ def text2svg(self):
...
+    CONFIG = {
+        "background_color": BLACK,
+    }
+
+    def __init__(self, text, **config):
+        Text.__init__(self, text, **config)
+        # self.text_backgrounds = self.gen_text_backgrounds(text)
+
+    def gen_text_backgrounds(self, text):
```
Comment (PgBiel) [inline manimlib/mobject/svg/text_mobject.py:326]: docs?

File: manimlib/mobject/svg/text_mobject.py [hunk 8]
```diff
@@ -314,86 +307,180 @@ def text2svg(self):
...
+        start_i = 0
+        for line_no in range(text_list.__len__()):
+            rect_counts = len(text_list[line_no])
+            text_backgrounds.add(*self[start_i:rect_counts])
+            start_i += 2 * rect_counts
+        text_backgrounds.set_color(self.background_color)
+
+        return text_backgrounds
+
+    def text2svg1(self):
```
Comment (PgBiel) [inline manimlib/mobject/svg/text_mobject.py:339]: docs?

File: manimlib/mobject/svg/text_mobject.py [hunk 9]
```diff
@@ -314,86 +307,180 @@ def text2svg(self):
...
+        self.lines[1].extend([self.alignment for _ in range(chars_lines_text_list.__len__())])
         VGroup.__init__(self, *[self.lines[0][i] for i in range(self.lines[0].__len__())], **config)
         self.move_to(np.array([0, 0, 0]))
-
-    def set_all_lines_alignment(self, alignment):
-        self.lines[1] = [alignment for _ in range(self.lines_list.__len__())]
+        if self.alignment:
+            self.set_all_lines_alignments(self.alignment)
+
+    def gen_chars(self, lines_str_list):
```
Comment (PgBiel) [inline manimlib/mobject/svg/text_mobject.py:447]: docs maybe

File: manimlib/mobject/svg/code_mobject.py [hunk 10]
```diff
@@ -1,52 +1,118 @@
...
+        It defines starting number for counting lines.
+    line_no_buff : :class:`float`, optional (default is 0.4)
+        It defines space between line numbers and displayed code.
+    style : :class:`str`, optional (default is 'vim')
+        It defines style type of displayed code. You can see names of styles from Code.styles_list.
+    language : :class:`str`, optional (default is None, Which mean it will automatically detects the language)
+        It defines the language name of given code.
+        For Knowing more available options visit https://pygments.org/docs/lexers/  for 'aliases or short names'
+    generate_html_file : :class:`bool`, optional (default is False)
+        It defines whether to generate code highlighted html to folder assets/codes/generated_html_files.
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:79]:
```suggestion
    tab_width : :class:`int`, optional
        Number of space characters corresponding to a tab character. Defaults to 3.
    line_spacing : :class:`float`, optional
        Amount of space between lines in relation to font size. Defaults to 0.3, which means 30% of font size.
    scale_factor : class:`float`, optional
        A number which scales displayed code. Defaults to 0.5.
    font : :class:`str`, optional
         The name of the text font to be used. Defaults to `"Monospac821 BT"`.
    stroke_width : class:`float`, optional
        Stroke width for text. 0 is recommended, and the default. 
    margin: class :`float`, optional
        Inner margin of text from the background. Defaults to 0.3.
    indentation_chars : :class:`str`, optional
        "Indentation chars" refers to the spaces/tabs at the beginning of a given code line. Defaults to `"    "`.
    background : :class:`str`, optional
        Defines the background's type. Currently supports only `"rectangle"` (default) and `"window"`.
    background_stroke_width : class:`float`, optional
        Defines the stroke width of the background. Defaults to 1.
    background_stroke_color : class:`str`, optional
        Defines the stroke color for the background. Defaults to `WHITE`.
    corner_radius : :class:`float`, optional
        Defines the corner radius for the background. Defaults to 0.2.
    insert_line_no : :class:`bool`, optional
        Defines whether line numbers should be inserted in displayed code. Defaults to `True`.
    line_no_from : :class:`int`, optional
        Defines the first line's number in the line count. Defaults to 1.
    line_no_buff : :class:`float`, optional
        Defines the spacing between line numbers and displayed code. Defaults to 0.4.
    style : :class:`str`, optional
        Defines the style type of displayed code. You can see possible names of styles in with :attr:`styles_list`. Defaults to `"vim"`.
    language : Optional[:class:`str`], optional
        Defines the language name of given code. If `None` (the default), the language will be automatically detected. For the list of possible options, visit https://pygments.org/docs/lexers/ and look for 'aliases or short names'.
    generate_html_file : :class:`bool`, optional (default is False)
        Defines whether to generate highlighted html code to the folder `assets/codes/generated_html_files`. Defaults to `False`.
```

File: manimlib/mobject/svg/code_mobject.py [hunk 11]
```diff
@@ -1,52 +1,118 @@
...
+    Attributes (It have all Parameters as Attributes and some extra as following)
+    ----------
+    background_mobject : :class:`~.VGroup`
+        To display background according to background type specified by background in Parameters.
+        VGroup with SurroundingRectangle() if background == "rectangle"
+        VGroup with RoundedRectangle() and Dot() for three buttons if background == "window"
+    line_numbers : :class:`~.Paragraph`
+        To display line numbers of displayed code if insert_line_no == True.
+    code : :class:`~.Paragraph`
+        To display highlighted code.
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:89]: what do you mean here by these "to display"s?

```suggestion
        The background displayed according to the background type specified by `background` in Parameters.
        This is a VGroup with :class:`~.SurroundingRectangle` if `background == "rectangle"`, and a VGroup with :class:`~.RoundedRectangle` and :class:`~.Dot` (three Dots for three buttons)) if `background == "window"`.
    line_numbers : :class:`~.Paragraph`
        The line numbers of displayed code if `insert_line_no == True`.
    code : :class:`~.Paragraph`
        The displayed highlighted code.
```

File: manimlib/mobject/svg/code_mobject.py [hunk 12]
```diff
@@ -245,6 +333,12 @@ def gen_code_json(self):
         # print(self.code_json)
 
     def correct_non_span(self, line_str):
+        """Function put text color to those strings that don't have one according to background_color of displayed code.
+        Arguments
+        ---------
+        line_str : :class:`str`
+            Takes a string to put color to it according to background_color of displayed code.
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:340]:
```suggestion
            Takes a string and colors it according to the :attr:`background_color` of displayed code.

```

File: manimlib/mobject/svg/code_mobject.py [hunk 13]
```diff
@@ -245,6 +333,12 @@ def gen_code_json(self):
         # print(self.code_json)
 
     def correct_non_span(self, line_str):
+        """Function put text color to those strings that don't have one according to background_color of displayed code.
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:336]:
```suggestion
        """Function that colors non-colored strings according to the :attr:`background_color` of displayed code.

```

File: manimlib/mobject/svg/code_mobject.py [hunk 14]
```diff
@@ -245,6 +333,12 @@ def gen_code_json(self):
         # print(self.code_json)
 
     def correct_non_span(self, line_str):
+        """Function put text color to those strings that don't have one according to background_color of displayed code.
+        Arguments
+        ---------
+        line_str : :class:`str`
+            Takes a string to put color to it according to background_color of displayed code.
+        """
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:341]: please document this return value
```suggestion

        Returns
        --------
        :class:`str`
        """
```

File: manimlib/mobject/svg/code_mobject.py [hunk 15]
```diff
@@ -272,8 +366,23 @@ def correct_non_span(self, line_str):
...
+        Code language name.
+    style : :class:`str`
+        Code style name.
+    insert_line_no : :class:`bool`
+        It defines whether insert line numbers in html file.
+    divstyles : :class:`str`
+        Some html css styles.
+    file_path : :class:`str`
+        Path of code file.
+       """
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:385]: wrong identation
```suggestion
    """

```

File: manimlib/mobject/svg/code_mobject.py [hunk 16]
```diff
@@ -272,8 +366,23 @@ def correct_non_span(self, line_str):
...
-def hilite_me(code, lexer, options, style, linenos, divstyles):
-    lexer = lexer or 'python'
+def hilite_me(code, language, style, insert_line_no, divstyles, file_path):
+    """Function to highlight code from string to html.
+    Arguments
+    ---------
+    code : :class:`str`
+        Code string.
+    language : :class:`str`
+        Code language name.
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:376]:
```suggestion
        The name of the programming language the given code was written in.

```

File: manimlib/mobject/svg/code_mobject.py [hunk 17]
```diff
@@ -272,8 +366,23 @@ def correct_non_span(self, line_str):
...
+    style : :class:`str`
+        Code style name.
+    insert_line_no : :class:`bool`
+        It defines whether insert line numbers in html file.
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:380]:
```suggestion
        Defines whether line numbers should be inserted in the html file.
```

File: manimlib/mobject/svg/text_mobject.py [hunk 18]
```diff
@@ -1,20 +1,39 @@
...
+from manimlib.mobject.geometry import Dot
 from manimlib.mobject.svg.svg_mobject import SVGMobject
 from manimlib.mobject.types.vectorized_mobject import VGroup
 from manimlib.utils.config_ops import digest_config
 
-
 TEXT_MOB_SCALE_FACTOR = 0.05
 
 
+def remove_invisible_chars(mobject):
```
Comment (PgBiel) [inline manimlib/mobject/svg/text_mobject.py:16]: maybe a little docstring could be useful here

File: manimlib/mobject/svg/code_mobject.py [hunk 19]
```diff
@@ -135,25 +196,40 @@ def ensure_valid_file(self):
                       self.file_name)
 
     def gen_line_numbers(self):
+        """Function to generate line_numbers.
+        Returns
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:200]: spacing
```suggestion

        Returns
```

File: manimlib/mobject/svg/code_mobject.py [hunk 20]
```diff
@@ -135,25 +196,40 @@ def ensure_valid_file(self):
...
-                            alignment="right", font=self.font, stroke_width=self.stroke_width).scale(self.scale_factor)
+                                 alignment="right", font=self.font, stroke_width=self.stroke_width).scale(
+            self.scale_factor)
+        for i in line_numbers:
+            i.set_color(self.default_color)
         return line_numbers
 
     def gen_colored_lines(self):
+        """Function to generate code.
+        Returns
```
Comment (PgBiel) [inline manimlib/mobject/svg/code_mobject.py:218]:
```suggestion

        Returns
```

File: manimlib/mobject/svg/text_mobject.py (hunk 6 above)
Comment (NavpreetDevpuri) [reply manimlib/mobject/svg/text_mobject.py:317]: I will. Firstly i supposed to learn from your suggestions on documentions of Code().

**Comment by NavpreetDevpuri** [comment]
Thanku @PgBiel for that.
Thats great !

File: manimlib/mobject/svg/text_mobject.py (hunk 6 above)
Comment (PgBiel) [reply manimlib/mobject/svg/text_mobject.py:317]: got it!

**Comment by PgBiel** [comment]
You're welcome! Make sure to take a look at each of the suggestions, and tell me if any questions remain.

**Comment by safinsingh** [comment]
@NavpreetDevpuri can you please bring this up to date with master?

**Comment by NavpreetDevpuri** [comment]
@safinsingh done https://github.com/ManimCommunity/manim/pull/198

**Comment by NavpreetDevpuri** [comment]
Is https://github.com/3b1b/manim no longer in developement ?

**Comment by Aathish04** [comment]
> Is https://github.com/3b1b/manim no longer in developement ?

It is still in development, see the shaders branch there. It's just that the de facto only developer there is 3b1b ;)

**Comment by NavpreetDevpuri** [comment]
alright understood, here we are developing this library further according to our needs and creativity and the main repo is used by 3b1b so he will be developing it according to his own requirement.

**Pull request closed without merging**